apps/passport-server/src/services/devconnectPretixSyncService.ts

```diff
@@ -46,126 +48,259 @@ interface EventData {
     } catch (e) {
+      logger(
+        `[DEVCONNECT PRETIX]: Encountered error when validating fetched data for ${this.organizer.id}: ${e}`
+      );
       this.rollbarService?.reportError(e);
-      logger("[DEVCONNECT PRETIX] Sync failed", e);
+
+      throw new Error("Data failed to validate", {
+        cause: errorCause("validating", this.organizer.id, e)
+      });
     }
-  }
 
-  public stop(): void {
-    if (this.timeout) {
-      clearTimeout(this.timeout);
+    try {
+      await this.save();
+    } catch (e) {
+      logger(
+        `[DEVCONNECT PRETIX]: Encountered error when saving data for ${this.organizer.id}: ${e}`
+      );
+      this.rollbarService?.reportError(e);
+
+      throw new Error("Data failed to validate", {
+        cause: errorCause("saving", this.organizer.id, e)
+      });
     }
+
+    return this.completeSuccess();
   }
 
   /**
-   * Download Pretix state, and apply a diff to our state so that it
-   * reflects the state in Pretix.
+   * Fetches data from Pretix. This will return a FetchOutcome, which
+   * can be either "rate-limited" or "complete". This is achieved by
+   * having two promises, one which resolves when we get rate-limited,
+   * and one which resolves when we finish fetching all available data.
+   *
+   * The first promise gets created on every run (that is, every time
+   * the function is called). It simply listens for the fetcher's pause
+   * event, and resolves if it is triggered.
+   *
+   * The second promise is potentially longer-lived, and survives across
+   * multiple calls. This is because it starts the fetching process,
+   * which may become paused due to rate-limiting.
+   *
+   * On each call, we reset the fetcher's count, which will un-pause the
+   * long-lived promise (if active). This causes it to resume, and will
+   * allow it to resolve its promise with the value of "complete".
    */
-  private async sync(pretixConfig: DevconnectPretixConfig): Promise<void> {
-    return traced(NAME, "sync", async (span) => {
-      span?.setAttribute("organizers_count", pretixConfig.organizers.length);
-
-      const syncStart = Date.now();
-
-      const organizerSyncPromises = []; // one per organizer
-      // Iterate over organizers and set up a sync promise for each
-      for (const organizer of pretixConfig.organizers) {
-        organizerSyncPromises.push(this.syncOrganizer(organizer));
-      }
-
-      try {
-        const results = await Promise.all(organizerSyncPromises);
-      } catch (e) {
-        logger(
-          "[DEVCONNECT PRETIX] Failed to save tickets for one or more events",
-          e
-        );
-        setError(e, span);
-        this.rollbarService?.reportError(e);
-      }
+  private async fetchData(): Promise<FetchOutcome> {
+    // This promise will resolve if we get rate-limited
+    const pausePromise = new Promise<"rate-limited">((resolve) => {
+      this.fetcher.once("pause", () => {
+        resolve("rate-limited");
```

Comment:
why not a void promise here?

Reply:
Same as the explanation above, we want to be able to distinguish between two promises that might resolve in a `Promise.any()` call